### PR TITLE
Add Subnet resource

### PR DESCRIPTION
**Pull request by brycahta**
**Issue:** [#489](https://github.com/aws-controllers-k8s/community/issues/489)

### Changes
* Adds create/delete Subnet functionality with smoke tests
* Adds terminal_codes for Subnet. Triggered when users try to create Subnet with malformed or invalid VpcId
  * related: attempting to delete VPC with Subnet attached will result in DependencyViolation which will requeue the delete with backoff (default behavior) until user resolves 
* Adds helm artifacts

### Testing
* Tested controller locally with `go run`  ✅
* `make kind-test` ✅

```
[gw2] [ 33%] PASSED tests/test_subnet.py::TestSubnet::test_terminal_condition
[gw3] [ 66%] PASSED tests/test_vpc.py::TestVpc::test_smoke
[gw0] [100%] PASSED tests/test_subnet.py::TestSubnet::test_crud

======================================================================= 3 passed in 54.60s =======================================================================
```

### CODE REVIEW

File: test/e2e/tests/test_subnet.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+test_resource_values = REPLACEMENT_VALUES.copy()
+
+
+@pytest.fixture(scope="module")
+def ec2_client():
+    return boto3.client("ec2")
+
+@pytest.fixture(scope="module")
+def vpc_resource():
```
Comment (brycahta) [inline test/e2e/tests/test_subnet.py:40]: My intent is to move this to a `helpers.py` file once I get more resources implemented. This way I can test VPC separately, then have another VPC (just one) to be used by other tests. Also, this should mitigate concerns with vpc quotas when testing locally.

File: apis/v1alpha1/types.go [hunk 2]
```diff
@@ -1586,10 +1708,13 @@ type LocalGatewayRouteTableVirtualInterfaceGroupAssociation struct {
 // Describes a local gateway virtual interface.
 type LocalGatewayVirtualInterface struct {
 	LocalAddress   *string `json:"localAddress,omitempty"`
+	LocalBGPASN    *int64  `json:"localBGPASN,omitempty"`
 	LocalGatewayID *string `json:"localGatewayID,omitempty"`
 	OwnerID        *string `json:"ownerID,omitempty"`
 	PeerAddress    *string `json:"peerAddress,omitempty"`
+	PeerBGPASN     *int64  `json:"peerBGPASN,omitempty"`
 	Tags           []*Tag  `json:"tags,omitempty"`
+	Vlan           *int64  `json:"vlan,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/types.go:1717]: Ooo found an initialism (I've been hunting)

File: pkg/resource/subnet/hooks.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"errors"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/ec2"
+)
+
+// isRequiredFieldsMissingFromInput returns True if CR is
+// missing required fields for reconciliation.
+func isRequiredFieldsMissingFromInput(r *resource) bool {
```
Comment (RedbackThomson) [inline pkg/resource/subnet/hooks.go:24]: super nit: `is` -> `are`?

File: test/e2e/tests/test_subnet.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+from acktest.k8s import resource as k8s
+from e2e import service_marker, CRD_GROUP, CRD_VERSION, load_ec2_resource
+from e2e.replacement_values import REPLACEMENT_VALUES
+
+RESOURCE_PLURAL = "subnets"
+
+CREATE_WAIT_AFTER_SECONDS = 10
+DELETE_WAIT_AFTER_SECONDS = 10
+
+test_resource_values = REPLACEMENT_VALUES.copy()
```
Comment (RedbackThomson) [inline test/e2e/tests/test_subnet.py:32]: I would recommend you put this inside the fixture, as you might want to re-use the fixture and not share the same replacement values.

File: test/e2e/tests/test_subnet.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        for subnet in subnets:
+            if subnet["SubnetId"] == subnet_id:
+                return subnet
+
+        return None
+
+    def subnet_exists(self, ec2_client, subnet_id: str) -> bool:
+        return self.get_subnet(ec2_client, subnet_id) is not None
+
+    def test_crud(self, ec2_client, vpc_resource):
```
Comment (RedbackThomson) [inline test/e2e/tests/test_subnet.py:86]: nit: This test doesn't test the "update" of CRUD

File: test/e2e/tests/test_subnet.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            resource_name, namespace="default",
+        )
+        k8s.create_custom_resource(ref, resource_data)
+        cr = k8s.wait_resource_consumed_by_controller(ref)
+
+        assert cr is not None
+        assert k8s.get_resource_exists(ref)
+
+        assert k8s.assert_condition_state_message(
+            ref, "ACK.Terminal", "True", "The vpc ID 'InvalidVpcId' does not exist"
```
Comment (RedbackThomson) [inline test/e2e/tests/test_subnet.py:151]: ++

**Comment by RedbackThomson** [comment]
Looks straightforward. Do your ReadMany changes in the generator not help here?

**Comment by brycahta** [comment]
> Looks straightforward. Do your ReadMany changes in the generator not help here?

Yeah, they do but only 1/2 are merged right now (working on revision) and it was unclear to me if I should wait until another release is cut with those features or not. Any suggestions?

File: pkg/resource/subnet/hooks.go (hunk 3 above)
Comment (brycahta) [reply pkg/resource/subnet/hooks.go:24]: resolved with auto-gen'd code

File: apis/v1alpha1/types.go (hunk 2 above)
Comment (brycahta) [reply apis/v1alpha1/types.go:1717]: https://github.com/aws-controllers-k8s/code-generator/pull/161

File: apis/v1alpha1/types.go [hunk 7]
```diff
@@ -1138,6 +1182,7 @@ type ImportVolumeTaskDetails struct {
 
 // Describes an instance.
 type Instance struct {
+	AmiLaunchIndex        *int64  `json:"amiLaunchIndex,omitempty"`
```
Comment (A-Hilaly) [inline apis/v1alpha1/types.go:1185]: not sure, but should this be `AMILaunchIndex`?

File: pkg/resource/subnet/hooks.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"errors"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/ec2"
+)
+
+// addIDToListRequest adds requested-resource VpcId to ListRequest. Return error to indicate to callers that the
+// resource is not yet created.
+func addIDToListRequest(r *resource, input *svcsdk.DescribeSubnetsInput) error {
+	if r.ko.Status.SubnetID == nil {
```
Comment (A-Hilaly) [inline pkg/resource/subnet/hooks.go:25]:
```suggestion
	if r.ko.Status.SubnetID == nil || *r.ko.Status.SubnetID == "" {
```

**Review by A-Hilaly — APPROVED** [review]
LGTM :+1:

File: pkg/resource/subnet/hooks.go (hunk 8 above)
Comment (brycahta) [reply pkg/resource/subnet/hooks.go:25]: good catch; will update copy-paste comment too

File: apis/v1alpha1/generator.yaml [hunk 9]
```diff
@@ -64,7 +65,13 @@ ignore:
...
-      sdk_read_many_pre_build_request:
-        template_path: hooks/vpc/sdk_read_many_pre_build_request.go.tpl
       sdk_read_many_post_build_request:
-        template_path: hooks/vpc/sdk_read_many_post_build_request.go.tpl
\ No newline at end of file
+        template_path: hooks/vpc/sdk_read_many_post_build_request.go.tpl
+  Subnet:
+    hooks:
+      sdk_read_many_post_build_request:
+        template_path: hooks/vpc/sdk_read_many_post_build_request.go.tpl
```
Comment (RedbackThomson) [inline apis/v1alpha1/generator.yaml:73]: nit: in `/vpc/` folder? Same logic?

**Review by RedbackThomson — APPROVED** [review]
/lgtm
Ship it!

**Pull request merged**